_drafts/hw_actors.md

```diff
@@ -0,0 +1,150 @@
+
+It is a common approach to use interrupt handlers as simple 'execution engine' in RTOS-less embedded systems. While it is better than loop-based solutions because of preemption, it may result in worse latencies in case when complex processing is performed in handlers directly. Many devices use interrupts by number of reasons, communication devices for example use them to signal transmission completion, new data receiving, errors and so on. When incoming data is processed in handlers, it blocks responses to other events, and this may be undesirable. Especially in case when communication protocol has timing constraints like USB where host periodically pings devices. So it is a good idea to split interrupt handlers to several independent parts and keep interrupt code as short as possible to be able to react to incoming events appropriately. Described approach may be considered as a way to do such split. Also, it simplifies timing and its use for timed events, and may help with processing events from multiple sources as well as utilization of multiple CPUs.
+
+Communicating actors may be easily put into another environment, like a host computer, for testing purposes. They do not need busy waiting, as in RTOS-less cases when the main program waits for data from interrupt handlers using ring buffers. This helps with both testability and power efficiency. Compared with threads, actors are less vulnerable to issues like unbounded priority inversion (but these issues are still possible!) and allow for better formal reasoning.
+
+However, the article would be incomplete without discussion about drawbacks. The most painful thing is the fact that actors may be deadlocked much easier than threads: any mistake in queue selection and the system will probably hang. Another disadvantage of the model is its requirement to redefine the task in terms of actors and messages, which may be not so easy. For example, just think about how to implement timeouts or waiting for a response before further message processing. Since the actor model relies on message allocations, it is less deterministic compared to fixed-memory solutions, like threads and semaphores. When message pools are exhausted, it is unclear how the handler should deal with it. In other words, design patterns for the actor model are less known and may be more complex than their threading-based counterparts are.
+
+In general, the actor model is a bad choice when you need some kind of global consensus or transactions (i.e., rollback of previous work when some actor at the end of the chain fails). So, while not all systems are suitable for the actor model, it may be a good choice for some and hardware implementation of parts of the model in widely available MCUs makes it yet more attractive.
```

Comment:
We generally like to have a stronger recommendation for what is being written about in the closing section. Try to give a compelling argument as to why what you're proposing is good, or has the ability to be good after some drawbacks are solved.